File: nim/day10.nim
import sequtils, strformat, sugar, tools, sets

type 
  TI = tuple[x, y: int]
  SI = seq[int]
  SSI = seq[SI]
  Visited = HashSet[TI]

proc readInput(filename: string): SSI =
  let file = open(filename, fmRead)
  defer: file.close()

  result = collect:
    for line in file.lines():
      line.toSeq().map(item => ord(item) - ord('0'))

func bfs(mat: SSI, pos: TI, visited: var Visited, target: int = 9): int =
  var localVisited = Visited()
  var queue: seq[TI]
  queue &= pos
  visited.incl(pos)
  localVisited.incl(pos)

  while queue.len() > 0:
    let current = queue.pop()
    if mat.fetchVal(current) == target:
      result += 1

    let (locs, nums) = nbrs(mat, current)
    for ind, val in locs:
      if nums[ind] == mat.fetchVal(current) + 1 and not localVisited.contains(val):
        queue &= val
        localVisited.incl(val)
        visited.incl(val)

func dfs(mat: SSI, pos: TI, visited: var Visited, targetHeight: int = 9): int =
  if mat.fetchVal(pos) == targetHeight:
    return 1

  visited.incl(pos)
  let (neighbors, _) = nbrs(mat, pos)
  for neighbor in neighbors:
    if neighbor notin visited and mat.fetchVal(neighbor) == mat.fetchVal(pos) + 1:
      result += dfs(mat, neighbor, visited)
  visited.excl(pos)
  
func solution(mat: SSI): (int, int) =
  var pt1_visited: Visited
  var pt2_visited: Visited

  for r_ind, r_val in mat:
    for c_ind, c_val in r_val:
      if c_val == 0: 
        if not pt1_visited.contains((r_ind, c_ind)):
          result[0] += bfs(mat, (r_ind, c_ind), pt1_visited)
        result[1] += dfs(mat, (r_ind, c_ind), pt2_visited)

let 
  mat = readInput("input/day10.txt")
  (pt1, pt2) = solution(mat)
echo &"Part 1: {pt1}\nPart 2: {pt2}"

File: cpp/day10.cpp
#include <print>
#include <unordered_set>
#include "tools.h"

constexpr size_t ROWS = 50;
constexpr size_t COLS = 50;

using aai = std::array<std::array<int, ROWS>, COLS>;
using spi = std::unordered_set<pi, pair_hash>;

auto parseFile() -> aai {
  std::ifstream file("input/day10.txt");
  assert(file);
  std::string line;

  aai arr = {{{0}}};

  int r_ind = 0;
  while (std::getline(file, line)) {
    for (size_t c_ind = 0; c_ind < line.size(); c_ind++) {
      arr[r_ind][c_ind] = line[c_ind] - '0';
    }
    r_ind++;
  }

  return arr;
}

auto bfs(const aai &mat, pi pos, spi &visited, int target) -> size_t {
  spi l_visited;
  std::vector<pi> queue;
  size_t ttl = 0;

  queue.emplace_back(pos);
  l_visited.insert(pos);

  while (queue.size() > 0) {
    pi current = queue.back();
    queue.pop_back();
    if (arrValue(std::span(mat), current) == target) ttl++;

    auto n = nbrs(std::span(mat), current, Direction::Udlr);
    for (size_t ind{0}; ind < n.indices.size(); ind++) {
      auto val = n.indices[ind];
      if ((n.vals[ind] == arrValue(std::span(mat), current) + 1) &&
          !l_visited.contains(val)) {
        queue.emplace_back(val);
        l_visited.insert(val);
        visited.insert(val);
      }
    }
  }

  return ttl;
}

auto dfs(const aai &mat, pi pos, spi &visited, int target) -> size_t {
  size_t ttl = 0;
  if (arrValue(std::span(mat), pos) == target) return 1;

  visited.insert(pos);
  auto nbr = nbrs(std::span(mat), pos, Direction::Udlr);
  for (const auto &n : nbr.indices) {
    if (!visited.contains(n) &&
        arrValue(std::span(mat), n) == arrValue(std::span(mat), pos) + 1) {
      ttl += dfs(mat, n, visited, 9);
    }
    visited.erase(pos);
  }

  return ttl;
}

auto solution(const aai &mat) -> void {
  size_t pt1{0}, pt2{0};
  spi pt1_visited;
  spi pt2_visited;

  int r_ind = 0;
  for (const auto &r_val : mat) {
    int c_ind = 0;
    for (const auto &c_val : r_val) {
      if (c_val == 0) {
        auto tmp = std::make_pair(r_ind, c_ind);
        if (!pt1_visited.contains(tmp)) {
          pt1 += bfs(mat, tmp, pt1_visited, 9);
        }
        pt2 += dfs(mat, tmp, pt2_visited, 9);
      }
      c_ind++;
    }
    r_ind++;
  }

  std::print("Part 1: {}\nPart 2: {}", pt1, pt2);
}

auto main() -> int {
  auto arr = parseFile();
  solution(arr);
}
